File: ios/CircularBuffer/CircularBufferModule.swift
import Foundation
import React

@objc(CircularBufferModule)
class CircularBufferModule: RCTEventEmitter {
    override static func moduleName() -> String! {
        return "CircularBufferModule"
    }

    override static func requiresMainQueueSetup() -> Bool {
        return false
    }

    // Camera errors (busy, session interrupted) surface asynchronously, well
    // after startBuffering's promise already resolved -- there's no pending
    // Promise left to reject, so they're emitted as a JS event instead.
    // useCircularBuffer.ts subscribes via circularBufferEvents and routes
    // this into the same error banner as a rejected saveClip().
    override func supportedEvents() -> [String]! {
        return ["CircularBufferError"]
    }

    private var moviesDir: URL {
        let documents = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        return documents.appendingPathComponent("PegueiClips", isDirectory: true)
    }

    @objc(startBuffering:resolver:rejecter:)
    func startBuffering(config: NSDictionary, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        guard
            let bufferSeconds = config["bufferSeconds"] as? Int,
            let width = config["width"] as? Int,
            let height = config["height"] as? Int,
            let fps = config["fps"] as? Int
        else {
            reject(CircularBufferErrorCode.encoderInitFailed, "Configuração de buffer inválida.", nil)
            return
        }

        let bufferConfig = BufferConfig(bufferSeconds: bufferSeconds, width: width, height: height, fps: fps)
        CameraEncoderController.shared.configureAndArm(config: bufferConfig, delegate: self)
        resolve(nil)
    }

    @objc(stopBuffering:rejecter:)
    func stopBuffering(resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        CameraEncoderController.shared.stop()
        resolve(nil)
    }

    @objc(startManualRecording:rejecter:)
    func startManualRecording(resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        // Resolving/rejecting a promise from a background thread is a
        // well-known crash source with the New Architecture (Fabric/
        // TurboModules, enabled in this project) -- CameraEncoderController's
        // callbacks fire from sessionQueue/DispatchQueue.global, not the main
        // thread. Hopping to the main queue before calling resolve/reject is
        // the standard, safe React Native convention. See DECISIONS.md.
        CameraEncoderController.shared.startManualRecording(
            moviesDir: moviesDir,
            onStarted: {
                DispatchQueue.main.async { resolve(nil) }
            },
            onFailed: { code, message in
                DispatchQueue.main.async { reject(code, message, nil) }
            }
        )
    }

    @objc(stopManualRecording:rejecter:)
    func stopManualRecording(resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        CameraEncoderController.shared.stopManualRecording(
            onSaved: { path, duration in
                DispatchQueue.main.async { resolve(["path": path, "durationSeconds": duration]) }
            },
            onFailed: { code, message in
                DispatchQueue.main.async { reject(code, message, nil) }
            }
        )
    }

    @objc(isBuffering:rejecter:)
    func isBuffering(resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        resolve(CameraEncoderController.shared.isBuffering)
    }

    @objc(setZoom:resolver:rejecter:)
    func setZoom(factor: NSNumber, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        CameraEncoderController.shared.setZoom(
            factor: factor.doubleValue,
            onSet: { applied in
                DispatchQueue.main.async { resolve(applied) }
            },
            onFailed: { code, message in
                DispatchQueue.main.async { reject(code, message, nil) }
            }
        )
    }

    @objc(getZoomInfo:rejecter:)
    func getZoomInfo(resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        CameraEncoderController.shared.zoomInfo(
            onResult: { minZoom, maxZoom, hasUltraWide in
                DispatchQueue.main.async {
                    resolve(["minZoom": minZoom, "maxZoom": maxZoom, "hasUltraWide": hasUltraWide])
                }
            },
            onFailed: { code, message in
                DispatchQueue.main.async { reject(code, message, nil) }
            }
        )
    }
}

extension CircularBufferModule: CameraEncoderControllerDelegate {
    func cameraEncoderController(_ controller: CameraEncoderController, didFailWithCode code: String, message: String) {
        sendEvent(withName: "CircularBufferError", body: ["code": code, "message": message])
    }
}

File: android/app/src/main/java/com/felipenunes/pegueiapp/circularbuffer/CircularBufferModule.kt
package com.felipenunes.pegueiapp.circularbuffer

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableNativeMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import java.io.File

class CircularBufferModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String = "CircularBufferModule"

    // Must match RNFS.DocumentDirectoryPath + "/PegueiClips" (files.ts CLIPS_DIR),
    // which is what the JS gallery reads from -- mirrors the iOS module's
    // documentDirectory/PegueiClips. Previously this pointed at
    // getExternalFilesDir(Movies), a different directory entirely, so saved
    // clips never showed up in the app's gallery.
    private fun moviesDir(): File {
        val dir = File(reactApplicationContext.filesDir, "PegueiClips")
        if (!dir.exists()) dir.mkdirs()
        return dir
    }

    /**
     * Camera errors (busy, disconnected) surface asynchronously, well after
     * startBuffering's promise already resolved -- there is no pending
     * Promise left to reject, so they're emitted as a JS event instead.
     * useCircularBuffer.ts subscribes via circularBufferEvents and routes
     * this into the same error banner as a rejected saveClip().
     */
    private fun emitError(code: String, message: String) {
        val params = WritableNativeMap()
        params.putString("code", code)
        params.putString("message", message)
        if (reactApplicationContext.hasActiveReactInstance()) {
            reactApplicationContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                .emit("CircularBufferError", params)
        }
    }

    @ReactMethod
    fun startBuffering(config: ReadableMap, promise: Promise) {
        try {
            val bufferConfig = BufferConfig(
                bufferSeconds = config.getInt("bufferSeconds"),
                width = config.getInt("width"),
                height = config.getInt("height"),
                fps = config.getInt("fps"),
            )
            CameraEncoderController.configureAndArm(
                reactApplicationContext,
                bufferConfig,
                object : CameraEncoderCallback {
                    override fun onError(code: String, message: String) {
                        emitError(code, message)
                    }
                },
            )
            promise.resolve(null)
        } catch (t: Throwable) {
            promise.reject(CircularBufferErrorCodes.ENCODER_INIT_FAILED, t.message, t)
        }
    }

    // NativeEventEmitter(NativeModules.CircularBufferModule) on the JS side
    // calls these on subscribe/unsubscribe; they're no-ops because events are
    // emitted straight through RCTDeviceEventEmitter above, not tracked here.
    @ReactMethod
    fun addListener(eventName: String) {}

    @ReactMethod
    fun removeListeners(count: Int) {}

    @ReactMethod
    fun stopBuffering(promise: Promise) {
        try {
            CameraEncoderController.stop()
            promise.resolve(null)
        } catch (t: Throwable) {
            promise.reject(CircularBufferErrorCodes.SAVE_FAILED, t.message, t)
        }
    }

    @ReactMethod
    fun startManualRecording(promise: Promise) {
        CameraEncoderController.startManualRecording(
            moviesDir(),
            object : StartRecordingCallback {
                override fun onStarted() {
                    promise.resolve(null)
                }

                override fun onFailed(code: String, message: String) {
                    promise.reject(code, message)
                }
            },
        )
    }

    @ReactMethod
    fun stopManualRecording(promise: Promise) {
        CameraEncoderController.stopManualRecording(
            object : SaveClipCallback {
                override fun onSaved(path: String, durationSeconds: Double) {
                    val result = WritableNativeMap()
                    result.putString("path", path)
                    result.putDouble("durationSeconds", durationSeconds)
                    promise.resolve(result)
                }

                override fun onFailed(code: String, message: String) {
                    promise.reject(code, message)
                }
            },
        )
    }

    @ReactMethod
    fun isBuffering(promise: Promise) {
        promise.resolve(CameraEncoderController.isBuffering())
    }

    @ReactMethod
    fun setZoom(factor: Double, promise: Promise) {
        try {
            val applied = CameraEncoderController.setZoom(factor.toFloat())
            promise.resolve(applied.toDouble())
        } catch (t: Throwable) {
            promise.reject(CircularBufferErrorCodes.NOT_BUFFERING, t.message, t)
        }
    }

    @ReactMethod
    fun getZoomInfo(promise: Promise) {
        val info = CameraEncoderController.getZoomInfo()
        if (info == null) {
            promise.reject(CircularBufferErrorCodes.NOT_BUFFERING, "O buffer não está ativo.")
            return
        }
        val result = WritableNativeMap()
        result.putDouble("minZoom", info.minZoom)
        result.putDouble("maxZoom", info.maxZoom)
        result.putBoolean("hasUltraWide", info.hasUltraWide)
        promise.resolve(result)
    }
}
